### PR TITLE
`FixedBufPool::next`

### PR DESCRIPTION
**Pull request by mzabaluev**
Add an async API to get buffers out of `FixedBufPool`. This is another way in which `FixedBufPool` will be more useful than `FixedBufRegistry`.

Also reorganize the shared data structures of `FixedBufPool` and `FixedBufRegistry` into separate modules for better readability and future code reuse.

### CODE REVIEW

**Review by Noah-Kennedy — APPROVED** [review]

**Comment by Noah-Kennedy** [comment]
ugh, I did not mean to remove the draft status

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [inline src/buf/fixed/plumbing/pool.rs:126]: Sounds like will_wake won't catch all duplicates. Is it possible this list will grow too large? And what's the danger of keeping an old waker that the scheduler had replaced in a subsequent poll?

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [inline src/buf/fixed/plumbing/pool.rs:149]: Is waking them all necessary because one or more may have been canceled?

**Review by FrankReh — COMMENTED** [review]
Nice.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:126]: > Sounds like will_wake won't catch all duplicates. Is it possible this list will grow too large?

It is theoretically, but I don't know a way to prevent this.

> And what's the danger of keeping an old waker that the scheduler had replaced in a subsequent poll?

I trust the `Waker` API to do the right thing in this case.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:149]: I don't get this question, sorry. All stored wakers are used up to wake any of the tasks that still remain, and `Waker::wake` is a no-op if its task has terminated. If a task has dropped the `Next` future (or otherwise stopped calling `poll_next`), it will get a spurious wakeup once, but I think this is normal in the async world.

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:149]: Oh, you got my question. You just couldn't hear my tone of voice of my body language. Like many of my questions, it came half from naiveté and half from just wanting to make conversation and wanting to double check and keep my brain thinking during the review.

It's the first time I've seen it in action but I agree it's necessary. It's maybe a small blemish on the async world where the wake call does not return the state of whether it found a Future to put into the `Ready` queue.

Thank you.

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:126]: Tokio still has an open issue about this. https://github.com/rust-lang/rust/issues/66281

How about keeping an optional clone in the future itself, to allow searching the `Vec` with the older one first, with simple address comparison? I'm afraid the `Vec`'s growth could be unbounded if the future were part of a select in a loop as long as Tokio's waker API doesn't guarantee will_wake doesn't have false negatives.

**Review by FrankReh — CHANGES_REQUESTED** [review]
The direction is great. Please consider my two outstanding questions. Thank you.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:25]: I don't think this requires an enum, can you sketch some code for that? Also, keeping the entry with an empty `Vec` value, even when all wakers for a given capacity are consumed, may be better than removing the `Vec` from the table and re-adding it later for another poll.

I thought about making this a hash table indexed per capacity. But then, the waker list(s) is not going to be populated unless the pool becomes exhausted, and when it does, we're anyway up for some latency, re-polling, spurious wakeups, etc.
So I would not like to make this logic more complex and branchy without having some performance data to understand whether this would make a significant difference.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:126]: > How about keeping an optional clone in the future itself, to allow searching the `Vec` with the older one first, with simple address comparison?

How could that work with cloned wakers that do not expose any underlying pointers in stable API?

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:126]: If you store a waker in the Future, don't you have a stable address you can compare against?

Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:126]: No. The reference returned by `cx.waker()` is not guaranteed to live longer than the poll call that the `Context` is passed into. Cloning the waker is possible, but what address comparison against other `Waker` instances is going to achieve then? The only API we have to avoid storing redundant wakers is `Waker::will_wake`. That's not guaranteed to always work, but I think in practice no executor will repeatedly produce wakers that can't be correlated to one another with `will_wake` when the task is repeatedly polled on a pending future. Correct me please if there are examples to the contrary.

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:126]: I'm sorry I don't see the problem yet with storing a cloned waker in the future when the cloned waker is also pushed onto the `Vec` so when poll is called again, before pushing a new cloned waker onto the `Vec`, the cloned waker held by the future is used to get the old waker's address and that address is searched in the `Vec` to see if a `Vec` location can be swapped out with a new clone. I have not seen any structures in the std library where a list of wakers is held. I'm prepared to write a demo to see if this idea would work, I don't see why it wouldn't - the waker is a function and it has an address that can be compared, but I haven't had time yet.

I know `will_wake` is used for the first best shot at deciding when a waker doesn't need to be saved because an equally valid waker is already held. But in all the cases I've seen, when `will_wake` returns false, it means a waker is replaced with a clone, not appended as a clone. I'm just trying to find the location in the `Vec` to perform the replacement by comparing pointers. Another approach, not relying on the waker address, would be to store a generation counter along with the waker in the `Vec` cell and keep the generation counter in the future, for searching the `Vec` on a subsequent poll.

Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs]: I have rewritten the wakeup implementation using `tokio::sync::Notify`. It's non-allocating, besides the synchronization objects created on demand for each capacity bucket.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs]: With the new implementation, wakeups do happen per capacity bucket, and a single task is awoken after each buffer check-in, if one is waiting for a buffer of this capacity.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [inline src/buf/fixed/plumbing/pool.rs:123]: The `Arc` is not necessary for the single-threaded pool, but I'm going to reuse this for the thread-safe implementation. `Notify` contains a mutex and uses atomic pointers for tracking waiters, so it's going to be slow regardless when we hit the waiting branch. Performance of `next` in the depleted pool case is not high on my current priority list for micro-optimizations.

File: tests/fixed_buf.rs
Comment (mzabaluev) [inline tests/fixed_buf.rs:167]: There should be a way to share `File` and socket objects between tasks, since the descriptor is already in a `SharedFd` internally.

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs]: Thanks for this clarification. I'll have to look at the code another time though to see how waking one and only one task is safe. Is the buffer ownership handed to the Future as part of being awoken, even before another poll? No need to answer, I'll look for the answer this weekend.

File: src/buf/fixed/pool.rs
Comment (FrankReh) [inline src/buf/fixed/pool.rs:284]: Maybe the comment could add that the permit may have been held even though there was no buffer, which is likely once the exhaustion case has gone away and then comes back.

**Review by FrankReh — APPROVED** [review]
I'm not going to check the unit tests tonight. But the use of Notify looks very nice. That is one powerful `waiting` mechanism `Tokio` has to offer. I ended up liking this implementation more than I thought I would because the Notify does all the heavy lifting of avoiding the pitfall of dropping an awake call. And as result, *all* the waiters don't have to be awoken to be on the safe side. The `Notify` implementation makes it safe to begin with, hence `notify_one` can be used very effectively.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:174]: Each buffer must be individually allocated since it's supplied to the constructor as a `Vec`.
I don't think io-uring requires the buffers to be aligned; tests pass with buffers of size 80, for example.

I put a `mem::drop` call in there to illustrate that the `Vec` is recovered from raw pointer and size specifically to be dropped and thus to get the buffer properly deallocated.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:27]: Yes, it's possible that a task calling `next` will consume the permit, not get a buffer from `try_next`, and continue waiting. It's also possible in other scenarios, for example when tasks are calling `next` and `try_next` on the same pool concurrently.

File: src/buf/fixed/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/pool.rs:285]: I lifted this from the code example on `Notified` :grin:

**Comment by mzabaluev** [comment]
@FrankReh
> But we are left without a simple/fast pool, that doesn't allocate, for the current-thread case.

While the pool is not exhausted, tasks call `next` at near zero cost with no allocations.

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:27]: Is that worth a comment somewhere, to let the reader know it has been considered, or perhaps that it is by design? When reading something new, it's nice to see that all the bases were covered by a design, before having to figure it out myself.

File: src/buf/fixed/plumbing/pool.rs
Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:174]: I'm not sure about this. If the constructor didn't do the individual allocators, it doesn't seem right that the drop would assume individual deallocations.

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:174]: At least some kernel ops require page aligned buffers. I don't recall which right now but why shouldn't the interface allow the caller to decide for themselves how the allocations are done? Does that make the API too onerous?

**Comment by FrankReh** [comment]
@mzabaluev 
> While the pool is not exhausted, tasks call `next` at near zero cost with no allocations.
That's a very good point. I think I overlooked that while digging into the weeds of the implementation.

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:174]: > If the constructor didn't do the individual allocators, it doesn't seem right that the drop would assume individual deallocations.

The constructor takes ownership of a sequence of `Vec` that have to be allocated using the global allocator, so drop deallocates these `Vec` objects. If the application wants a more precise control over the allocations, perhaps it should consider its own implementation of a `FixedBuffers` collection (when we get around to make the trait public).

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:27]: https://github.com/tokio-rs/tokio-uring/pull/199/commits/d83eacb9a484be4cb9567a9977ea117fa833fc56

File: src/buf/fixed/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/pool.rs]: https://github.com/tokio-rs/tokio-uring/pull/199/commits/d83eacb9a484be4cb9567a9977ea117fa833fc56

File: src/buf/fixed/plumbing/pool.rs
Comment (mzabaluev) [reply src/buf/fixed/plumbing/pool.rs:125]: https://github.com/tokio-rs/tokio-uring/pull/199/commits/091cb4de5548a6311b2d351c6c358ad1d90461d8

**Comment by ollie-etl** [comment]
@Noah-Kennedy @mzabaluev Is something blocking this? It looks like all points have been resolved

**Comment by Noah-Kennedy** [comment]
You don't need my approval here. I'm alright with @FrankReh signing off on this and merging it whenever he feels it is ready.

File: src/buf/fixed/plumbing/pool.rs
Comment (ollie-etl) [reply src/buf/fixed/plumbing/pool.rs:125]: I'd just like to clarify my understanding here: once a notify has been registered for a capacity, then every time a buffer of that capacity is commited, notify_one() is called (because the notify_next_by_cap entry is never removed)?

**Comment by oliverbunting** [comment]
> @mzabaluev 
> > While the pool is not exhausted, tasks call `next` at near zero cost with no allocations.
> 
> That's a very good point. I think I overlooked that while digging into the weeds of the implementation.
> 

Is this not a common case though. I.e, to keep some pipeline of tasks fully utilized, do work whenever there is a buffer available?

**Comment by mzabaluev** [comment]
> Is this not a common case though. I.e, to keep some pipeline of tasks fully utilized, do work whenever there is a buffer available?

Depends if the task creation is only throttled by availability of the buffers.

File: src/buf/fixed/plumbing/pool.rs
Comment (ollie-etl) [reply src/buf/fixed/plumbing/pool.rs:125]: I think this could be improved by keeping a `(usize, Notify)` instead, to explicitly keep a count for the number of waiting tasks, and removing  the entry from the hashset once count == zero. That way, you only pay the `SeqCst` cost of calling `notify()` when you need to.

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:125]: Not saying this is needed, I will let @mzabaluev set the path on this one. But I would prefer something that had a little hysteresis to it. Like removing the Notify if the pool size reached something like 10 buffers again. (But I have no idea why I would pick 10. Just something larger than 0.) But I haven't tried to design anything around this idea so feel free to disregard entirely.

Comment (ollie-etl) [reply src/buf/fixed/plumbing/pool.rs:125]: My concern is that with the current design, once `try_next()` has failed once in the `next()` method. every checking from then on for that buffer size is going to issue a `SeqCst` memory barrier with the call to `notify_one()`. Thats quite a penalty. Its unavoidable with the surrent Notify, which is designed for `Send` operations, but we could at least limit it to when there is actually a consumer waiting.

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:125]: @ollie-etl With your idea, I don't see how the usize field is decremented properly? In the case of a future being cancelled, there is no future to do the work of updating the count (from my understanding).

Comment (ollie-etl) [reply src/buf/fixed/plumbing/pool.rs:125]: You'd need to return a `Next` struct wrapper, which decrements on drop

Comment (FrankReh) [reply src/buf/fixed/plumbing/pool.rs:125]: Oh, that's much more like @mzabaluev 's original idea I think, of returning a future that does work when it is polled, and does work when it is dropped. But using Notify.

**Comment by FrankReh** [comment]
I'm going to merge later tonight so it makes it into the next release. This is some really great work that we want in the repo but I had to do a manual merge on four files because code moved and code changed at the same time.

Actually the condensed diff that this page makes available isn't too bad to read. Most of the changes are straight forward. 

I guess the drop code, and the intra-doc links are where there could be problems so another pair of eyes or two or three wouldn't hurt. It compiles cleanly, and there is a nice unit test that passes. But rarely do our unit tests check that things were dropped properly so that's why looking at the drop again would make sense.

But also, nothing we can't change yet again with more iterations.

**Comment by FrankReh** [comment]
Thank you @mzabaluev .

**Pull request merged**
